app/assets/src/components/common/MetadataInput.jsx

```diff
@@ -0,0 +1,65 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { isArray } from "lodash";
+
+import Input from "~/components/ui/controls/Input";
+import Dropdown from "~/components/ui/controls/dropdowns/Dropdown";
+import DateInput from "~/components/ui/controls/DateInput";
+
+class MetadataInput extends React.Component {
+  render() {
+    const { value, onChange, onSave, metadataType, className } = this.props;
+
+    if (isArray(metadataType.options)) {
+      const options = metadataType.options.map(option => ({
+        text: option,
+        value: option
+      }));
+      return (
+        <Dropdown
+          fluid
+          floating
+          options={options}
+          onChange={val => onChange(metadataType.name, val, true)}
+          value={value}
+          className={className}
+        />
+      );
+    } else if (metadataType.dataType == "date") {
+      return (
+        <DateInput
+          className={className}
+          onChange={val => onChange(metadataType.name, val, true)}
+          value={value}
+        />
+      );
+    } else {
+      return (
+        <Input
+          className={className}
+          onChange={val => onChange(metadataType.name, val)}
+          onBlur={() => onSave && onSave(metadataType.name)}
+          value={value}
+          type={metadataType.dataType === "number" ? "number" : "text"}
+        />
+      );
+    }
+  }
+}
+
+MetadataInput.propTypes = {
+  className: PropTypes.string,
+  value: PropTypes.any,
+  metadataType: PropTypes.shape({
+    key: PropTypes.string,
+    dataType: PropTypes.oneOf(["number", "text", "date"]),
+    options: PropTypes.arrayOf(PropTypes.string)
+  }),
+  // First parameter is the metadataType NAME
+  // Third optional parameter is whether to immediately save. false means "wait for onSave to fire".
```

Comment:
I'm confused by the `Third optional parameter` comment? which one is that?

Reply:
Let me fix the comment.

We needed some way to tell the parent when to save.
For `<Inputs>`, we want the parent to save when the user blurs, not while the user is typing. But we still need to tell the parent about changes when the user is typing.

Reply:
This is referring to the `true` in `onChange(metadataType.name, val ,true)`